homeassistant/components/subaru/config_flow.py

```diff
@@ -0,0 +1,170 @@
+"""Config flow for Subaru integration."""
+from datetime import datetime
+import logging
+
+from subarulink import Controller as SubaruAPI, SubaruException
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import (
+    CONF_DEVICE_ID,
+    CONF_PASSWORD,
+    CONF_PIN,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import callback
+from homeassistant.helpers import aiohttp_client, config_validation as cv
+
+from .const import (
+    CONF_HARD_POLL_INTERVAL,
+    DEFAULT_HARD_POLL_INTERVAL,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    MIN_HARD_POLL_INTERVAL,
+    MIN_SCAN_INTERVAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+        vol.Required(CONF_PIN): str,
+    }
+)
+
+
+@callback
+def configured_instances(hass):
+    """Return a set of configured Subaru instances."""
+    return {entry.title for entry in hass.config_entries.async_entries(DOMAIN)}
+
+
+class SubaruConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Subaru."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_import(self, import_config):
+        """Import a config entry from configuration.yaml."""
+        return await self.async_step_user(import_config)
+
+    async def async_step_user(self, user_input=None):
+        """Handle the start of the config flow."""
+
+        if not user_input:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={},
+                description_placeholders={},
+            )
+
+        if user_input[CONF_USERNAME] in configured_instances(self.hass):
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "identifier_exists"},
+                description_placeholders={},
+            )
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "connection_error"},
+                description_placeholders={},
+            )
+        except InvalidAuth:
+            return self.async_show_form(
+                step_id="user",
+                data_schema=DATA_SCHEMA,
+                errors={"base": "invalid_credentials"},
+                description_placeholders={},
```

Comment:
When showing the form again, it'll probably be a better user experience to pre-populate the fields with the previous values.

Reply:
Form is now pre-populated for both invalid password and invalid PIN errors.